### DISCUSSION TITLE
Nothing happens when installing the ep_setup.exe

### DISCUSSION THREAD

**Discussion by HenrikoMagnifico** [valinet/ExplorerPatcher#566] [Q&A]
I double clicked the setup installer, it asked for elevation. Clicked "Ok" and then the screen went gray for a second. Then nothing changed. Everything's like usual in Windows 11, what did I do wrong?

Running the latest 2021-12, cumulative update for Windows 11 for x64-based systems (KB5008215)

**Comment by valinet** [comment]
Kill `explorer.exe` and start it manually.

**Comment by Andre-10** [comment]
Do you see "Properties" when right clicking the Taskbar?

**Comment by null-von-sushi** [comment]
For what it's worth, it also did nothing for me. Right clicking the taskbar just gives me "Taskbar properties", which opens the new settings. I tried restarting explorer manually, did not fix it. I am using an ARM based system though...

**Reply by valinet** [reply]
Yeah, no chances anything works at all on ARM64 as is distributed. The architecture matters! The things this kind of application does are dependent on a lot of things cooperating together, including the CPU architecture. Besides, it's never been tested on an ARM64 machine, as I do not own one; also, it would have a much real chance of working if it would be compiled for ARM64, not for x64 and emulated via WOW. As is, it can't work. This application comes in the form of a dynamic link library that gets injected into `explorer.exe`. On an ARM64 machine, `explorer.exe` is an ARM64 executable, while ExplorerPatcher is compiled and distributed for amd64 and IA-32. You cannot injected a module of a different architecture than the executable. Really, much better results if this were compiled for ARM64.

**Reply by null-von-sushi** [reply]
That's fair enough. I wasn't aware on exactly  how it's implemented. Surprisingly, a lot of stuff (save for drivers) works well on Micorosft's compatibility layer.  Not the end of the world, I just wanted to mention my results :)

**Reply by SRTRVTH** [reply]
@valinet seeing as you don't have an ARM64 machine, I'm assuming support for those devices aren't coming any time soon? Would love to run this on a Pro X, but regardless thank you for all of the work you do! Running beautifully on another Windows laptop.

**Reply by valinet** [reply]
@SRTRVTH Yeah, indeed, not coming anytime soon, or at least not from me directly. The source code is public, everyone is free to port it.
@null-von-sushi This is the closest thing to a driver that still runs in user space, I’d say, so yeah, can’t really work unless a native port is done.

**Reply by 7coil** [reply]
Fun times - I've wasted a few hours of my life into compiling an ARM64 version, and I think I may be getting _somewhere_.
I've retained funchook and am attempting to use capstone instead of distorm for whatever this program does.

I can't seem to understand how the setup program works, and have broken my Shell Experience Host :(

If anyone has a fork they want me to try out, feel free to let me try.

**Reply by valinet** [reply]
Use Microsoft Detours which works in place, and is already included. There is some define at the beginning of dllmain.c, afair.

**Reply by valinet** [reply]
Not dllmain.c, hooking.h, there is a define: #define HOW_TO_HOOK HOOK_WITH_FUNCHOOK; change that to #define HOW_TO_HOOK HOOK_WITH_DETOURS. Make sure to compile Microsoft Detours beforehand. I have done the groundwork for this previously when the application was much smaller and I thought I might be able to come up with a working ARM64 build without testing (in fact, I was chatting with someone and he was testing, but had no means for debugging). Yeah, not really a productive working environment, so I gave up on the idea.

**Reply by valinet** [reply]
And it’s not wasted time, I call it learning experience. Yeah, it takes time, that’s what computers are about, everything takes ages to do.

I can explain in more detail how  ep_setup works, what exactly have you had trouble wrapping your head around with?

**Reply by 7coil** [reply]
Hello again!
I've tried switching to Microsoft Detours, but I just get nothing.

![image](https://user-images.githubusercontent.com/13886925/148466067-58c5f201-5e2f-4adc-a11f-e619682ab943.png)

I (think) it's being loaded:
![image](https://user-images.githubusercontent.com/13886925/148466095-4157f12c-39c2-40c9-ac52-32ec60a5f56a.png)

If you're wondering how _i screwed it up_, have a gander: https://github.com/7coil/ExplorerPatcher/tree/detours

**Reply by 7coil** [reply]
I also just edited references in the setup file from AMD64 to ARM64, which makes me feel that the setup works properly

**Reply by valinet** [reply]
> makes me feel that the setup works properly

It’s a Turing machine, you don’t need to “feel”, we have all the instruments to actually confirm something. If the solution file generates ExplorerPatcher.amd64.dll, that has to be changed as well. Anyway, idk, the setup probably has to be modified: on ARM64, you would actually need 3 DLLs:
* ARM64 version to inject explorer and native processes
* amd64 version to inject x86 64-bit apps running under emulation (if EP is used as shell extension)
* IA-32 version to inject x86 32-bit apps running under emulation (if EP is used as shell extension)

Also, idk if the calling convention used by Windows APIs on ARM64 is different than the one used by the standard. Have to check it out. If it is different, many function prototypes are improper as they are, they would have to be decorated with the calling convention as well (using “WINAPI”, which on amd64 is “__fastcall”). The reason these are missing in the first place is, at first, I wasn’t that aware of this aspect (EP is a giant learning process) and because on amd64 the Windows API uses fastcall while the standard uses fastcall as well. On IA-32, they actually differ, like, Windows API is fastcall while the standard uses cdecl as far as I remember, so for the functions included there, those are decorated.

Also, some patches have to be manually edited, especially the ones handpicking class members from class pointers (obtained by calls to GetWindowLongPtrW, for example). For example, extracting the clock object instance from the CTray object. Or TrayUI from CTray. Those are highly dependent on the object layout generated by the compiler. They 99.99% do not match what the amd64 compiler did. The disassembly for the ARM64 version of explorer.exe has to be studied. Unfortunately, IDA Free only works with amd64 binaries. Maybe try Ghidra…?

This patch is not trivial at all. New architectures don’t come for free, or as the touted magic pills where everything is solved magically. Or at least for these kind of apps. Yeah, web pages generally work just fine…

**Reply by 7coil** [reply]
I have builds working for all three architecture; I'll take a gander to see what I can **bodge** tomorrow.

Also it's a "feeling" because even with the tools, I don't really understand what I'm doing - I'm just a monkey that bought into WOA because I'm an idiot.

**Reply by msikma** [reply]
It does nothing for me either, and I'm also on Windows 11 for ARM. Curious to see what you come up with.

**Reply by Amrsatrio** [reply]
Here's my latest progress on getting this to work on ARM64:
https://github.com/valinet/ExplorerPatcher/discussions/138#discussioncomment-10024144
